# Week 3. Please note that I have stuffed everything in the same notebook, but I have used Markdown to help you identify the three different segments of the assignment.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


# Part 1 starts here

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [3]:
raw_data = pd.read_html(url)

In [4]:
df = raw_data[0]

In [5]:
df = df[df.Borough != "Not assigned"]

In [6]:
df.head()

Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [7]:
df.shape

(103, 3)

In [10]:
df['Postal code'].value_counts().sum()

103

### The total number of rows in our data frame is 103. We can also see the all values in "Postal code" column are unique using the "value_counts()" because the length here is also 103

In [11]:
df['Neighborhood'].isnull().sum()

0

### Also there is no "Not assigned" row in the "Neighborhood" column

In [12]:
df.shape

(103, 3)

## ================================================

# Part 2 Starts here

In [13]:
#Using the CSV option
csv_url = "http://cocl.us/Geospatial_data"
latlon_df = pd.read_csv(csv_url)

In [14]:
latlon_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [15]:
latlon_df.columns

Index(['Postal Code', 'Latitude', 'Longitude'], dtype='object')

In [16]:
latlon_df.rename(columns={'Postal Code':'Postal code'},inplace=True)

In [17]:
can_df = pd.merge(df, latlon_df, on='Postal code')

In [18]:
can_df.head(15)

Postal code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   

                                         Neighborhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                          Regent Park / Harbourfront  43.654260 -79.360636  
3                   Lawrence Manor / Lawrence Heights  43.718518 -79.464763  
4        Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
5                                    Islington Avenue  43.667856 -79.532242  
6                                     Malvern / Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                    Parkview Hill / Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park / Princess Gardens / Martin Gr...  43.650943 -79.554724  
12           Rouge Hill / Port Union / Highland Creek  43.784535 -79.160497  
13                                          Don Mills  43.725900 -79.340923  
14                                   Woodbine Heights  43.695344 -79.318389

In [19]:
can_df.shape

(103, 5)

## ============================================================

# Part 3 Starts here

In [20]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(can_df['Borough'].unique()),
        can_df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### slicing the original dataframe and create a new dataframe of Toronto data.

### Only working with boroughs containing the word "Toronto"

In [21]:
toronto_df = can_df[can_df['Borough'].str.contains("Toronto")]

In [22]:
toronto_df.reset_index(drop=True,inplace=True)

In [24]:
toronto_df.head(10)

Postal code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park / Harbourfront   
1         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2         M5B  Downtown Toronto                      Garden District, Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   
5         M5E  Downtown Toronto                                   Berczy Park   
6         M5G  Downtown Toronto                            Central Bay Street   
7         M6G  Downtown Toronto                                      Christie   
8         M5H  Downtown Toronto                    Richmond / Adelaide / King   
9         M6H      West Toronto                 Dufferin / Dovercourt Village   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031  
5  43.644771 -79.373306  
6  43.657952 -79.387383  
7  43.669542 -79.422564  
8  43.650571 -79.384568  
9  43.669005 -79.442259

In [26]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toront are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toront are 43.6534817, -79.3839347.


In [27]:
# create map of Manhattan using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [28]:
CLIENT_ID = '04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK' # your Foursquare ID
CLIENT_SECRET = 'VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4' # your Foursquare Secret
VERSION = '20180605'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK
CLIENT_SECRET:VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's latitude and longitude values.

In [30]:
neighborhood_latitude = toronto_df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_df.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park / Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [31]:
# type your answer here
LIMIT = 100
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # disp




'https://api.foursquare.com/v2/venues/explore?&client_id=04O0JVTWOOFL3ZO4SMR3T3C3YUFARKYEGB5L0UT5DKCU3NAK&client_secret=VJDTRRG5DWMGXZ0ZOXMVAIZ53U3VE0JQVUTQYE2SZPJ3EDQ4&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

In [32]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5eac939ab9a3890028b11393'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 48,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

In [33]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [34]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/home/saad/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
3  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

And how many venues were returned by Foursquare?

In [35]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

48 venues were returned by Foursquare.


##  Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [37]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [38]:
# type your answer here

toronto_venues = getNearbyVenues(names=toronto_df['Neighborhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )



Regent Park / Harbourfront
Queen's Park / Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
The Danforth West / Riverdale
Toronto Dominion Centre / Design Exchange
Brockton / Parkdale Village / Exhibition Place
India Bazaar / The Beaches West
Commerce Court / Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park / The Junction South
North Toronto West
The Annex / North Midtown / Yorkville
Parkdale / Roncesvalles
Davisville
University of Toronto / Harbord
Runnymede / Swansea
Moore Park / Summerhill East
Kensington Market / Chinatown / Grange Park
Summerhill West / Rathnelly / South Hill / Forest Hill SE / Deer Park
CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport
Roseda

#### Let's check the size of the resulting dataframe

In [39]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park / Harbourfront               43.65426              -79.360636   
1  Regent Park / Harbourfront               43.65426              -79.360636   
2  Regent Park / Harbourfront               43.65426              -79.360636   
3  Regent Park / Harbourfront               43.65426              -79.360636   
4  Regent Park / Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

Let's check how many venues were returned for each neighborhood

In [41]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton / Parkdale Village / Exhibition Place                         23   
Business reply mail Processing CentrE                                  18   
CN Tower / King and Spadina / Railway Lands / H...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             36   
Davisville North                                                        8   
Dufferin / Dovercourt Village                                          19   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         23   
India Bazaar / The Beaches West                                        20   
Kensington Market / Chinatown / Grange Park                            55   
Lawrence Park                                                           4   
Little Portugal / Trinity                                              41   
Moore Park / Summerhill East                                            1   
North Toronto West                                                     20   
Parkdale / Roncesvalles                                                13   
Queen's Park / Ontario Provincial Government                           38   
Regent Park / Harbourfront                                             48   
Richmond / Adelaide / King                                             94   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede / Swansea                                                    39   
St. James Town                                                         77   
St. James Town / Cabbagetown                                           46   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West / Rathnelly / South Hill / Fore...                     17   
The Annex / North Midtown / Yorkville                                  22   
The Beaches                                                             4   
The Danforth West / Riverdale                                          43   
Toronto Dominion Centre / Design Exchange                             100   
University of Toronto / Harbord                                        36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton / Parkdale Village / Exhibition Place                          23   
Business reply mail Processing CentrE                                   18   
CN Tower / King and Spadina / Railway Lands / H...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              36   
Davisville North                    

#### Let's find out how many unique categories can be curated from all the returned venues

In [43]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


## 3. Analyze Each Neighborhood

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot
toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               1   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0        0     0                          0            0   
1        0     0                          0            0   
2        0     0                          0            0   
3        0     0                          0            0   
4        0     0                          0            0   

In [45]:
toronto_onehot.shape

(1623, 231)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [47]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio   Airport  \
0                                        Berczy Park     0.000000  0.000000   
1     Brockton / Parkdale Village / Exhibition Place     0.000000  0.000000   
2              Business reply mail Processing CentrE     0.055556  0.000000   
3  CN Tower / King and Spadina / Railway Lands / ...     0.000000  0.058824   
4                                 Central Bay Street     0.015873  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.000000      0.000000        0.000000         0.000000   
2            0.000000      0.000000        0.000000         0.000000   
3            0.058824      0.058824        0.117647         0.176471   
4            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000                  0.0           0.0       0.0     0.017544   
1          0.000000                  0.0           0.0       0.0     0.000000   
2          0.000000                  0.0           0.0       0.0     0.000000   
3          0.117647                  0.0           0.0       0.0     0.000000   
4          0.000000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.015873                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0       0.000000   0.017544         0.0    0.017544  0.035088   0.0  0.000000   
1       0.000000   0.000000         0.0    0.000000  0.043478   0.0  0.043478   
2       0.055556   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
3       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.058824   
4       0.000000   0.000000         0.0    0.000000  0.000000   0.0  0.031746   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.017544  0.017544              0.0  0.035088   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Boat or Ferry  Bookstore  \
0         0.0                 0.0  0.017544       0.000000        0.0   
1         0.0                 0.0  0.000000       0.000000        0.0   
2         0.0                 0.0  0.000000       0.000000        0.0   
3         0.0                 0.0  0.000000       0.058824        0.0   
4         0.0                 0.0  0.000000       0.000000        0.0   

   Boutique  Brazilian Restaurant  Breakfast Spot   Brewery  Bubble Tea Shop  \
0  0.000000                   0.0        0.017544  0.000000         0.000000   
1  0.000000                   0.0        0.086957  0.000000         0.000000   
2  0.000000                   0.0        0.000000  0.055556         0.000000   
3  0.058824                   0.0        0.000000  0.000000         0.000000   
4  0.000000                   0.0        0.000000  0.000000         0.031746   

   Building  Burger Joint  Burrito Place  Bus Line  Butcher      Café  \
0       0.0      0.000000       0.000000       0.0      0.0  0.035088   
1       0.0      0.000000       0.043478       0.0      0.0  0.130435   
2       0.0      0.000000       0.055556       0.0      0.0  0.000000   
3       

In [48]:
toronto_grouped.shape

(39, 231)

#### Let's print each neighborhood along with the top 5 most common venues

In [50]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3          Restaurant  0.04
4  Seafood Restaurant  0.04


----Brockton / Parkdale Village / Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4   Grocery Store  0.04


----Business reply mail Processing CentrE----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2       Pizza Place  0.06
3        Comic Shop  0.06
4  Recording Studio  0.06


----CN Tower / King and Spadina / Railway Lands / Harbourfront West / Bathurst Quay / South Niagara / Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3          Boutique  0.06
4  Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1  Italian Restaurant  0.08
2              

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [51]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [54]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing CentrE           Yoga Studio   
3  CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar           Cheese Shop   
1           Coffee Shop             Nightclub        Breakfast Spot   
2         Auto Workshop                  Park            Comic Shop   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                       Café                Bakery   
1             Pet Store                    Stadium          Intersection   
2           Pizza Place           Recording Studio         Burrito Place   
3              Boutique        Rental Car Location                 Plane   
4        Ice Cream Shop  Middle Eastern Restaurant                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant                Bistro              Jazz Club  
1                Bakery    Italian Restaurant             Restaurant  
2            Restaurant               Brewery     Light Rail Station  
3         Boat or Ferry       Harbor / Marina                    Bar  
4       Thai Restaurant          Burger Joint    Fried Chicken Joint

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [55]:
# set number of clusters
kclusters = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [56]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal code           Borough                                  Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park / Harbourfront   
1         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2         M5B  Downtown Toronto                      Garden District, Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M4E      East Toronto                                   The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               1           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1                  Café   
4  43.676357 -79.293031               0                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery                   Pub   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café        Cosmetics Shop   
3           Coffee Shop          Cocktail Bar   American Restaurant   
4     Health Food Store                   Pub         Women's Store   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0        Breakfast Spot                  Café                      Theater   
1         Burrito Place              Beer Bar           Italian Restaurant   
2            Restaurant   Japanese Restaurant           Italian Restaurant   
3             Gastropub                 Hotel                   Restaurant   
4          Dance Studio     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0    Mexican Restaurant                 Restaurant         Chocolate Shop  
1             Juice Bar             Sandwich Place           Burger Joint  
2       Bubble Tea Shop  Middle Eastern Restaurant       Ramen Restaurant  
3                   Gym         Italian Restaurant       Department Store  
4            Donut Shop           Doner Restaurant                Dog Run

Finally, let's visualize the resulting clusters

In [57]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Cluster 1

In [58]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  East Toronto               0                 Trail     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                   Pub         Women's Store          Dance Studio   

  6th Most Common Venue        7th Most Common Venue 8th Most Common Venue  \
4     Electronics Store  Eastern European Restaurant            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
4      Doner Restaurant                Dog Run

Cluster 2

In [59]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               1           Coffee Shop   
1   Downtown Toronto               1           Coffee Shop   
2   Downtown Toronto               1        Clothing Store   
3   Downtown Toronto               1                  Café   
5   Downtown Toronto               1           Coffee Shop   
6   Downtown Toronto               1           Coffee Shop   
7   Downtown Toronto               1         Grocery Store   
8   Downtown Toronto               1           Coffee Shop   
9       West Toronto               1                Bakery   
10  Downtown Toronto               1           Coffee Shop   
11      West Toronto               1                   Bar   
12      East Toronto               1      Greek Restaurant   
13  Downtown Toronto               1           Coffee Shop   
14      West Toronto               1                  Café   
15      East Toronto               1        Sandwich Place   
16  Downtown Toronto               1           Coffee Shop   
17      East Toronto               1                  Café   
20   Central Toronto               1                  Park   
22      West Toronto               1    Mexican Restaurant   
23   Central Toronto               1           Coffee Shop   
24   Central Toronto               1                  Café   
25      West Toronto               1             Gift Shop   
26   Central Toronto               1          Dessert Shop   
27  Downtown Toronto               1                  Café   
28      West Toronto               1           Pizza Place   
30  Downtown Toronto               1                  Café   
31   Central Toronto               1                   Pub   
32  Downtown Toronto               1       Airport Service   
34  Downtown Toronto               1           Coffee Shop   
35  Downtown Toronto               1           Coffee Shop   
36  Downtown Toronto               1           Coffee Shop   
37  Downtown Toronto               1           Coffee Shop   
38      East Toronto               1           Yoga Studio   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                   Park                Bakery                          Pub   
1       Sushi Restaurant                 Diner                  Yoga Studio   
2            Coffee Shop                  Café               Cosmetics Shop   
3            Coffee Shop          Cocktail Bar          American Restaurant   
5           Cocktail Bar              Beer Bar                  Cheese Shop   
6                   Café    Italian Restaurant               Sandwich Place   
7                   Café                  Park                  Candy Store   
8                   Café            Restaurant              Thai Restaurant   
9               Pharmacy           Supermarket                  Music Venue   
10              Aquarium                 Hotel                         Café   
11            Restaurant           Men's Store        Vietnamese Restaurant   
12           Coffee Shop    Italian Restaurant               Ice Cream Shop   
13                 Hotel                  Café                   Restaurant   
14           Coffee Shop             Nightclub               Breakfast Spot   
15  Fast Food Restaurant         Movie Theater                          Pub   
16                  Café            Restaurant                        Hotel   
17           Coffee Shop             Gastropub                       Bakery   
20      Department Store                   Gym               Breakfast Spot   
22                  Café       Thai Restaurant         Fast Food Restaurant   
23        Clothing Store                  Park           Salon / Barbershop   
24        Sandwich Place           Coffee Shop                  Pizza Place   
25      Cuban Restaurant        Breakfast Spot  Eastern European Restaurant   
26        Sandwich Place           Pizza Place                  Coffee Shop   
27      

Cluster 3

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18   Central Toronto               2           Swim School   
21   Central Toronto               2                  Park   
33  Downtown Toronto               2                  Park   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
18                  Park  Construction & Landscaping              Bus Line   
21         Jewelry Store                       Trail              Bus Line   
33            Playground                       Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
18         Women's Store  Ethiopian Restaurant            Electronics Store   
21      Sushi Restaurant         Deli / Bodega            Electronics Store   
33          Dance Studio     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Eastern European Restaurant            Donut Shop       Doner Restaurant  
21  Eastern European Restaurant            Donut Shop       Doner Restaurant  
33                   Donut Shop      Doner Restaurant                Dog Run

Cluster 4

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19         Women's Store         Deli / Bodega           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
19  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19            Donut Shop      Doner Restaurant                Dog Run

Cluster 5

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29         Women's Store         Deli / Bodega  Ethiopian Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
29     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29      Doner Restaurant               Dog Run    Distribution Center

# End of Notebook